Source/Core/Common/DirectoryIterator.h

```diff
@@ -0,0 +1,127 @@
+// This file is under the public domain.
+
+#pragma once
+
+#include <iterator>
+#include <memory>
+#include <string>
+
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <dirent.h>
+#endif
+
+#include "Common/CommonTypes.h"
+
+namespace Common
+{
+// This class is designed for speed over convenience.
+// You may want to consider using FileSearch.h instead.
+class DirectoryEntry final
+{
+public:
+#ifdef _WIN32
+  using CharType = wchar_t;
+  using StringType = std::wstring;
+#else
+  using CharType = char;
+  using StringType = std::string;
+#endif
+
+  DirectoryEntry();
+  explicit DirectoryEntry(const StringType& path);
+  ~DirectoryEntry();
+
+  DirectoryEntry(const DirectoryEntry& other) = delete;
+  DirectoryEntry(DirectoryEntry&& other);
+  DirectoryEntry& operator=(const DirectoryEntry& other) = delete;
+  DirectoryEntry& operator=(DirectoryEntry&& other);
+
+  void Advance();
+  bool IsValid() const;
+
+  // Returns the name, not the full path. Only call if IsValid returns true.
+  const CharType* GetName() const;
+
+  // Runs faster than File::IsDirectory on many systems. Only call if the following are true:
+  // 1. IsValid returns true, and
+  // 2. The OS is Windows or the path reference passed to the constructor is still valid
+  // (Note that ToOSEncoding passes through references untouched on non-Windows systems.)
+  bool IsDirectory() const;
+
+  // Runs faster than File::GetSize on Windows. Only call if the following are true:
+  // 1. IsValid returns true, and
+  // 2. The OS is Windows or the path reference passed to the constructor is still valid
+  // (Note that ToOSEncoding passes through references untouched on non-Windows systems.)
+  u64 GetSize() const;
+
+private:
+  void Close();
+  void Move(DirectoryEntry&& other);
+
+  void SkipSpecialName();
+
+#ifdef _WIN32
+  HANDLE m_handle = INVALID_HANDLE_VALUE;

```

Comment:
Could you get rid of enough calls to Close() if you turned Close() into a deleter and the handle into a unique_ptr for it to be worth it in regards to the more complicated calls to the traversal methods?

Reply:
We only have two calls to `Close`, so I don't think it's worth it.